File: Laba1/Massiv.cpp
#include "head.h"
//МАССИВ
void createMas(Numbers& A, int C) {
    A.num = new string[C];
    A.size = 0; // изначально элементов нет
    A.C = C;
}
void addMasAtInd(Numbers& A, int ind, string znach) {// Добавить элемент по индексу
    if (ind < 0 || ind >= A.C) {
        cout << "Индекс вне диапазона емкости массива." << endl;
        return;
    }
    if (ind >= A.size) {
        for (int i = A.size; i < ind; i++) {
            A.num[i] = "";
        }
        A.size = ind + 1;
    }
    A.num[ind] = znach;
}
void addMasAtEnd(Numbers& A, string znach) {// Добавить элемент в конец
    if (A.size >= A.C) {
        cout << "Массив заполнен." << endl;
        return;
    }
    A.num[A.size] = znach;
    A.size++;
}
void poiskMasPoInd(Numbers& A, int ind) {// Получить элемент по индексу (вывод)
    if (ind < 0 || ind >= A.size) {
        cout << "Выход за границы массива." << endl;
        return;
    }
    cout << A.num[ind] << endl;
}
void deleteMasPoInd(Numbers& A, int ind) {// Удалить элемент по индексу
    if (ind < 0 || ind >= A.size) {
        return; 
    }
    for (int i = ind; i < A.size - 1; i++) {
        A.num[i] = A.num[i + 1];
    }
    A.size--; // Уменьшаем размер
}
void ZamenaMas(Numbers& A, int ind, string znach) {// Заменить элемент по индексу
    if (ind < 0 || ind >= A.size) {
        cout << "Нельзя заменить то что не существует." << endl;
        return;
    }
    A.num[ind] = znach;
}
void RazmerMas(Numbers& A) {// Получить размер массива
    cout << A.size << endl;
}
void readMas(Numbers& A) {// Вывести все элементы массива
    if (A.size == 0) {
        cout << "Массив пустой, Ваша честь." << endl;
        return;
    }
    for (int i = 0; i < A.size; i++) {
        cout << A.num[i] << " ";
    }
    cout << endl;
}

File: Laba1(Go)/Massiv.go
package main

import (
	"fmt"
)

// Numbers - структура массива
type Numbers struct {
	num  []string
	size int
	C    int
}
// createMas - создание массива
func createMas(C int) Numbers {
	return Numbers{
		num:  make([]string, C),
		size: 0,
		C:    C,
	}
}
// addMasAtInd - добавить элемент по индексу
func addMasAtInd(A *Numbers, ind int, znach string) {
	if ind < 0 || ind >= A.C {
		fmt.Println("Индекс вне диапазона емкости массива.")
		return
	}
	if ind >= A.size {
		for i := A.size; i < ind; i++ {
			A.num[i] = ""
		}
		A.size = ind + 1
	}
	A.num[ind] = znach
}
// addMasAtEnd - добавить элемент в конец
func addMasAtEnd(A *Numbers, znach string) {
	if A.size >= A.C {
		fmt.Println("Массив заполнен.")
		return
	}
	A.num[A.size] = znach
	A.size++
}
// poiskMasPoInd - получить элемент по индексу (вывод)
func poiskMasPoInd(A *Numbers, ind int) {
	if ind < 0 || ind >= A.size {
		fmt.Println("Выход за границы массива.")
		return
	}
	fmt.Println(A.num[ind])
}
// deleteMasPoInd - удалить элемент по индексу
func deleteMasPoInd(A *Numbers, ind int) {
	if ind < 0 || ind >= A.size {
		return
	}
	// Сдвигаем элементы влево
	for i := ind; i < A.size-1; i++ {
		A.num[i] = A.num[i+1]
	}
	// Очищаем последний элемент
	A.num[A.size-1] = ""
	A.size-- // Уменьшаем размер
}
// ZamenaMas - заменить элемент по индексу
func ZamenaMas(A *Numbers, ind int, znach string) {
	if ind < 0 || ind >= A.size {
		fmt.Println("Нельзя заменить то что не существует.")
		return
	}
	A.num[ind] = znach
}
// RazmerMas - получить размер массива
func RazmerMas(A *Numbers) {
	fmt.Println(A.size)
}
// readMas - вывести все элементы массива
func readMas(A *Numbers) {
	if A.size == 0 {
		fmt.Println("Массив пустой, Ваша честь.")
		return
	}
	for i := 0; i < A.size; i++ {
		fmt.Print(A.num[i], " ")
	}
	fmt.Println()
}
// getSize - получить размер массива (возвращает значение)
func getSize(A *Numbers) int {
	return A.size
}
// getCapacity - получить емкость массива
func getCapacity(A *Numbers) int {
	return A.C
}
